pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,359 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pulumi/pulumi/pkg/apitype"
+	"github.com/pulumi/pulumi/pkg/backend"
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/engine"
+	"github.com/pulumi/pulumi/pkg/operations"
+	"github.com/pulumi/pulumi/pkg/resource/config"
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/util/fsutil"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/pkg/util/result"
+	"github.com/pulumi/pulumi/pkg/workspace"
+	uuid "github.com/satori/go.uuid"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type lockContent struct {
```

Comment:
It would be ideal if we could document what our locking strategy here is and why we picked the scheme we did.  I'll summarize what I believe it is, but you can help me understand if I got it wrong or not.

To support locking, we add a new folder in the bookkeeping dir, named `locks`.  This directory will hold files named like the following:

`<stack-name>.<uuid>.json`

The UUID is assigned randomly per lockable backend and (and changes anytime a new lockable backend is created, which in the case of the CLI is more or less once per invocation).

To lock a stack, we create a file in this locks folder.  Because the UUID ensures the file name is unique, even on S3, we get read-after-write consistency.

Because each invocation of the CLI gets its own UUID assigned, as part of locking we have to ensure that there is only a single file in the locks folder for a given stack before knowing we have done our locking.  We confirm this by listing all the objects in the bucket after we have created our lock file to ensure we are the only copy.  If so, we know we got the lock and if not there was a race and we lost.

Do I have the right of this?

If I understand this correctly, I'm a little nervous about the approach.  From the [S3 dock](https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel):

> Amazon S3 achieves high availability by replicating data across multiple servers within Amazon's data centers. If a PUT request is successful, your data is safely stored. However, information about the changes must replicate across Amazon S3, which can take some time, and so you might observe the following behaviors:
>
> - A process writes a new object to Amazon S3 and immediately lists keys within its bucket. Until the change is fully propagated, the object might not appear in the list.

To my mind, this means that it is possible for there to be a case where P1 and P2 are racing to lock a stack.  Both processes write a lock file and then list the contents of the bucket to ensure they are the only copy.  In this case, I believe that it is possible for that the lists P1 and P2 do only end up seeing the lock they wrote, if the locks the others wrote have not propagated across the entire system.  In that case, I believe that both P1 and P2 would believe they own the lock and things would go bad after that.

@bigkraig do you believe my understanding here is correct?  If so - I'm not really sure if we can use this approach...

Reply:
@ellismg that is correct. perhaps we should state that if this isn't good enough for people, the Pulumi service doesn't have this problem? :)

Reply:
I suppose that's why Terraform needs DynamoDB to handle concurrent access on S3 state backend.

Reply:
Indeed, DynamoDB will provide strongly consistent read which will help us solve it. However, S3 only have the read after write consistency when the object first created.